kafka/source/pkg/reconciler/kafkasource.go

```diff
@@ -206,6 +219,188 @@ func checkResourcesStatus(src *v1alpha1.KafkaSource) error {
+	scaledObjectResourceInterface := r.DynamicClientSet.Resource(gvr).Namespace(namespace)
+	if scaledObjectResourceInterface == nil {
+		return nil, fmt.Errorf("unable to create dynamic client for ScaledObject")
+	}
+	scaledObjectUnstr, err := r.generateKedaScaledObjectUnstructured(ctx, ra, src)
+	if err != nil {
+		return nil, err
+	}
+	created, err := scaledObjectResourceInterface.Create(scaledObjectUnstr, metav1.CreateOptions{})
+	if err != nil {
+		logger.Error("Failed to create ScaledObject so going to do update", zap.Error(err))
+		//fmt.Printf("Doing update as failed to create ScaledObject: %s \n", err)
+		// will replace - https://github.com/kubernetes/client-go/blob/master/examples/dynamic-create-update-delete-deployment/main.go
+		// doing kubectl "replace" https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency
+		// first get resourceVersion
+		existing, err := scaledObjectResourceInterface.Get(name, metav1.GetOptions{})
+		if err != nil {
+			logger.Error("Failed to create ScaledObject:", zap.Error(err))
+			return nil, err
+		}
+		resourceVersion := existing.GetResourceVersion()
+		scaledObjectUnstr.SetResourceVersion(resourceVersion)
+		updated, err := scaledObjectResourceInterface.Update(scaledObjectUnstr, metav1.UpdateOptions{})
+		if err != nil {
+			logger.Error("Update failed to create ScaledObject", zap.Error(err))
+			return nil, err
+		} else {
+			logger.Info("Update success", zap.Any("updated", updated))
+			return updated, nil
+		}
+	}
+	return created, nil
+}
+
+const (
+	kedaAutoscalingAnnotationClass       = "keda.autoscaling.knative.dev"
+	kedaCooldownPeriodAnnodationKey      = kedaAutoscalingAnnotationClass + "/cooldownPeriod"
+	kedaPollingIntervalAnnodationKey     = kedaAutoscalingAnnotationClass + "/pollingInterval"
+	kedaTriggerLagThresholdAnnodationKey = kedaAutoscalingAnnotationClass + "/trigger.lagThreshold"
+)
+
+func convertMapKeyToInt32(dict map[string]string, key string, logger *zap.Logger) *int32 {
+	val, ok := dict[key]
+	if !ok {
+		return nil
+	}
+	i, err := strconv.ParseInt(val, 10, 32)
+	if err != nil {
+		logger.Error("Expected annotation value to be integer but got "+val, zap.Any("annotations key", key))
+		return nil
+	}
+	i32 := int32(i)
+	return &i32
+}
+
+func (r *Reconciler) readScalingAnnotations(ctx context.Context, src *v1alpha1.KafkaSource) {
+	logger := logging.FromContext(ctx).Desugar()
+	meta := src.GetObjectMeta()
+	annotations := meta.GetAnnotations()
+	if annotations != nil {
+		scalingClass := annotations[autoscaling.ClassAnnotationKey]
+		r.minReplicaCount = convertMapKeyToInt32(annotations, autoscaling.MinScaleAnnotationKey, logger)
+		r.maxReplicaCount = convertMapKeyToInt32(annotations, autoscaling.MaxScaleAnnotationKey, logger)
+		if scalingClass == kedaAutoscalingAnnotationClass {
+			r.cooldownPeriod = convertMapKeyToInt32(annotations, kedaCooldownPeriodAnnodationKey, logger)
```

Comment:
as mentioned above, this seems wrong. The parameters of a particular source shouldn't be saved in the reconciler object

Reply:
the mental model i have is that Kafka Source controller is responsible for managing all aspect so Kafka Source CR and that includes managing scaling CR and that scaling si the default mode. Non-scaling that is not using KEDA (or HPA) is legacy in future?

Reply:
hmm I think scaling shouldn't be the default mode... It will depend on what users want when they create their KafkaSource objects.
A user can create a KafkaSource A with no scaling annotation, so they don't want that to scale.
Or a user may want to create a KafkaSource B with the scaling annotation that they want to be scalable....